aerleon/lib/proxmox.py

```diff
@@ -0,0 +1,658 @@
+                ValueOption(config, log_level_forward=Proxmox._LOG_LEVELS),
+                ValueOption(config, policy_forward=list(Proxmox._TERM.ACTIONS_MAP.values())),
+            ],
+        },
+    }
+
+    def __init__(self, pol: Policy, exp_info: int):
+        self.proxmox_policies = []
+        super().__init__(pol, exp_info)
+
+    def _BuildTokens(self) -> Tuple[Set[str], Dict[str, Set[str]]]:
+        """returns the list of DSL + YAML supplementary tokens supported (proxmox-firewall specific)"""
+        supported_tokens, supported_sub_tokens = super()._BuildTokens()
+        supported_tokens |= {
+            # proxmox firewall only supports setting the source interface
+            'source_interface',
+            # proxmox firewall supports icmp type + icmp code
+            'icmp_code',
+        }
+        supported_sub_tokens.update({'option': set(self._TERM.LOG_LEVELS_MAP_OPTIONS.keys())})
+        return supported_tokens, supported_sub_tokens
+
+    def _TranslatePolicy(self, pol: policy.Policy, exp_info: int) -> None:
+        filter_zones = []
+        global_policy_config = ProxmoxConfigDataClass()
+        for header, terms in pol.filters:
+            self.filter_options = header.FilterOptions(self._PLATFORM)
+
+            if len(self.filter_options) < 2:
+                raise UnsupportedFilterOptionError(
+                    "missing options, zone and direction are mandatory"
+                )
+
+            filter_zone = self.filter_options[0]
+            filter_direction = self.filter_options[1]
+
+            if filter_zone not in self._BY_ZONE.keys():
+                raise UnsupportedFilterOptionError("unknown zone")
+
+            if filter_zones and filter_zone not in filter_zones:
+                raise ZoneMismatchError("cannot mix zone types in one policy")
+            filter_zones.append(filter_zone)
+
+            if filter_direction not in self._BY_ZONE[filter_zone]["supported_directions"]:
+                raise UnsupportedFilterOptionError(
+                    "direction "
+                    + filter_direction
+                    + " not supported for zone type "
+                    + filter_zone,
+                )
+
+            filter_config = ProxmoxConfigDataClass()
+            available_zone_options = self._BY_ZONE[filter_zone]['supported_options'](filter_config)
+            for t in self.filter_options[2:]:
+                ingested_options = []
+                for o in available_zone_options:
+                    ingested_options.append(o.ingest(t))
+                if not any(ingested_options):
+                    raise UnsupportedFilterOptionError("incorrect filter option directive %s" % t)
+            incomplete_options = list(filter(lambda o: not o.complete(), available_zone_options))
+            if incomplete_options:
+                raise UnsupportedFilterOptionError(
+                    "missing or incorrect value for filter option(s) %s", incomplete_options
```

Comment:
Can incomplete_options ever be an object? If so then we should probably wrap this in str.